content/2016-07-05-pythran-practice.md

```diff
@@ -0,0 +1,433 @@
+Title: Pythran learn by practicing
+Date: 2016-07-04
+Category: examples
+Lang: en
+Authors: pbrunet
+Summary: Show Pythran usage on multiple kernel with its strength and its weakness
+
+Pythran presentation hold on Université Lyon1 on Monday 27th June together with a Julia presentation.
+
+You can get slides shown by serge-sans-paille [here](http://serge-sans-paille.github.io/talks/python-perf-2016-06-27.html).
+
+After this presentation, some examples were proposed to see Pythran usage.
+
+# First example : Euler project N°14
+
+## Problem
+
+The following iterative sequence is defined for the set of positive integers
+n -> n/2 (n is even)
+n -> 3n + 1 (n is odd)
+
+Using the rule above and starting with 13, we generate the following sequence:
+13 -> 40 -> 20 -> 10 -> 5 -> 16 -> 8 -> 4 -> 2 -> 1
+
+It can be seen that this sequence (starting at 13 and fiinishing at 1) contains
+10 terms.
+Although it has not been proved yet (Collatz Problem), it is thought that all
+starting numbers finish at 1.
+
+Which starting number, under one million, produces the longest chain?
+
+NOTE: Once the chain starts the terms are allowed to go above one million.
+
+## Running the script
+
+We need to get the maximum length of each values.
+
+```python
+>>> def run():
+>>>     max_k, max_v = 1, 0
+>>>     for i in xrange(1000000):
+>>>         new_len = collatz_chain_length(i)
+>>>         if new_len > max_v:
+>>>             max_k, max_v = i, new_len
+>>>     return max_k
+```
+
+Now, we just need to compute the length of the chain for a given value.
+
+## A recursive solution
+
+Searching on inernet, it is a common solution like [here](https://www.lucaswillems.com/fr/articles/40/project-euler-14-solution-python)
+or [here](https://github.com/nayuki/Project-Euler-solutions/blob/master/python/p014.py)
+
+```python
+>>> def collatz_chain_length(x):
+>>>     if x not in collatz_cache:
+>>>         if x % 2 == 0:
+>>>             y = x // 2
+>>>         else:
+>>>             y = x * 3 + 1
+>>>     collatz_cache[x] = collatz_chain_length(y) + 1
+>>>     return collatz_cache[x]
+```
+
+This recursive version need a global variable to cache data as Python is limited by the recursion depth.
+As Pythran can't handle non constant globals, we will pass this variable as a second argument
+
+Now time it
+
+```bash
+python -c "import prob_14; print prob_14.run()"
+python -m timeit -s "import prob_14" "prob_14.run()"
+```
+
+```bash
+837799
+10 loops, best of 3: 1.26 msec per loop
+```
+<!-- It is 189 msec with the globals variable but I don't understand why it is so fast... -->
```

Comment:
just so you know : accessing a global variable in cpython is slower than accessing a local variable